packages/webdriver/src/node/bidi.ts

```diff
@@ -0,0 +1,109 @@
+        ])).flat()
+        // If the host resolves to a single IP address
+        // then it does not make sense to try additional candidates
+        // as the web socket DNS resolver would do extactly the same
+        if (candidateIps.length > 1) {
+            const hostnameMapper = (ip: string) => webSocketUrl.replace(parsedUrl.hostname, ip)
+            candidateUrls.push(...candidateIps.map(hostnameMapper))
+        }
+    }
+    return candidateUrls
+}
+
+interface ConnectionResult {
+    index: number
+    ws: WebSocket
+    isConnected: boolean
+    errorMessage?: string
+}
+
+/**
+ * Connect to a websocket
+ * @param candidateUrls - list of websocket urls to try
+ * @returns true if the connection was successful
+ */
+export async function connectWebsocket(candidateUrls: string[], _?: unknown): Promise<WebSocket | undefined> {
+    const websockets: WebSocket[] = candidateUrls.map((candidateUrl) => {
+        log.debug(`Attempt to connect to webSocketUrl ${candidateUrl}`)
+        return new WebSocket(candidateUrl)
+    })
+
+    const wsConnectPromises: Promise<ConnectionResult>[] = websockets.map((ws, index) => {
+        return new Promise<ConnectionResult>((resolve) => {
+            ws.once('open', () => resolve({ ws, isConnected: true, index }))
+            ws.once('error', (err) => {
+                log.debug(`Could not connect to Bidi protocol at ${candidateUrls[index]}: ${err.message}`)
+                resolve({ ws, isConnected: false, errorMessage: err.message, index })
+            })
+        })
+    })
+
+    const connectionTimeoutPromise = new Promise<undefined>((resolve) => {
+        setTimeout(() => {
+            log.error(`Could not connect to Bidi protocol of any candidate url: "${candidateUrls.join('", "')}"`)
+            return resolve(undefined)
+        }, CONNECTION_TIMEOUT)
+    })
+
+    const ws = await Promise.race([
+        firstResolved(wsConnectPromises),
+        connectionTimeoutPromise,
+    ])
+
+    const socketsToCleanup = ws ? websockets.filter((_, index) => ws.index !== index) : websockets
+    for (const socket of socketsToCleanup) {
+        socket.removeAllListeners()
+        socket.terminate()
+    }
+
+    if (ws?.isConnected) {
+        log.info(`Connected to Bidi protocol at ${candidateUrls[ws.index]}`)
+        return ws.ws
+    }
+
+    return undefined
+}
+
+/**
+ * Race the promises and return the first resolved promise that shows a successful connection
+ * @param promises - list of promises to race
+ * @returns the first resolved promise
+ */
+function firstResolved (promises: Promise<ConnectionResult>[]): Promise<ConnectionResult> {
+    // Race the wrapped promises
+    return Promise.race(promises).then(result => {
+        if (result.isConnected) {
+            return result
+        }
+
+        // If the first result was a rejection, race the remaining promises
+        return firstResolved(promises.filter((_, index) => index !== result.index))
```

Comment:
not sure if this logic would work as expected if there are more than two promises in the list. I would rather only consider filtering for promises that are in pending state. Check, for example, https://dev.to/devcrafter91/elegant-way-to-check-if-a-promise-is-pending-577g#:~:text=The%20trick,Promise%20%7B%20undefined%20%7D'%20when%20finished.

Reply:
Since the size of the array never change we can rely on the fact that a certain index can point to a specific promise.

Reply:
Lets check an example:

- We have 3 promises in the list
- The second promise has failed and returned undefined
- We now race between promise 1 and 3
- The third promise has failed and returned undefined
- We now race between promise 2 and 1
- Promise 2 is currently fulfilled, so it wins the race with undefined immediately
- We now race between promise 1 and 3
- we continue going to the recursive call infinite number of times if none of promises connects

Another example of undefined behaviour for Promise.race is when we only have one item in the initial list, which fails to connect and then we call the recursion with an empty list

Reply:
You are right, I had a logic mistake in there. I pushed a commit with a fix. Good catch 👌